### ISSUE TITLE
Pattern discard allowed for union case that takes no data

### ISSUE THREAD

**Issue by NinoFloris** [dotnet/fsharp#13851]
Match discards aren't type checked against case

```fs
    type X = X 

    let test() =
        match X with
        | X _ -> () // No error X doesn't have arguments
```

**Comment by vzarytovskii** [comment]
Duplicate of https://github.com/dotnet/fsharp/issues/10204 and https://github.com/dotnet/fsharp/issues/1592, which are not considered bugs, according to @dsyme.

**Comment by edgarfgp** [comment]
It always triggers me that when refactoring a DU my code still still complex where I’m using _ in pattern matching 😟

**Comment by dsyme** [comment]
Things have changed a bit since 2016, and we now are more inclined to emit new informationals and warnings, protected by a `/langversion` switch.  

So I'd definitely be ok with emitting at least an informational, and possibly a warning for this case.

**Comment by NinoFloris** [comment]
Was this originally missed or was this intended as a source compatibility feature? (being able to drop data in select scenarios)

Do informational diags pop up in IDEs? I wouldn't mind a warning tbh.

**Comment by dsyme** [comment]
> Was this originally missed or was this intended as a source compatibility feature? (being able to drop data in select scenarios)

It crept in somewhere along the way

> Do informational diags pop up in IDEs? 

Yes they show, but very light decorations

**Comment by abelbraaksma** [comment]
@edgarfgp great you’re trying to fix that old bug I reported 6yrs back in #1592. Once this is done, we should probably change the docs accordingly, see suggestion here on this issue: https://github.com/dotnet/docs/issues/20858.

Btw, if we make this informational (not a warning), surely a langversion switch isn’t necessary?

**Comment by edgarfgp** [comment]
@abelbraaksma Thanks for raising this years ago , This was confusing to me while I was learning . So I decided fix it :) 

>  Once this is done, we should probably change the docs accordingly, see suggestion here on this issue: [dotnet/docs#20858](https://github.com/dotnet/docs/issues/20858).

Yeah that is good idea. will update the docs if this gets merged 


> Btw, if we make this informational (not a warning), surely a langversion switch isn’t necessary?

I thought the same while I was implementing this and IMO a warning (no informational) will be more explicit and will help us remove those unnecessary wildcards. Also, it seems the informational warning is shown, but with very light decorations.

**Comment by T-Gro** [comment]
I agree that **warning** + language switch is better in making this more visible.
Info-warnings are use very sparsely, typically only when deprecating language features...

**Comment by vzarytovskii** [comment]
> I agree that **warning** + language switch is better in making this more visible.
> Info-warnings are use very sparsely, typically only when deprecating language features...

Agree, I think it should be a warning behind language version. It should also be easy to add a quick fix to VS (and other IDEs) which will be removing unnecessary discards.

**Comment by abelbraaksma** [comment]
Awesome to see this fixed!
